### ISSUE TITLE
Support multiple screens

### ISSUE THREAD

**Issue by bwikbs** [flutter-tizen/engine#94]
- Support ``FlutterEngineGroup`` to tizen 
https://flutter.dev/docs/development/add-to-app/multiple-flutters

- Run sample at tizen device 
https://github.com/flutter/samples/tree/master/add_to_app/multiple_flutters

**Comment by bwikbs** [comment]
Currently, My goal is to get the following screen with  a new ``FlutterEngineGroup`` API. Because I think this can satisfy the requirements what I understand.
![dump_screen](https://user-images.githubusercontent.com/4053790/120123215-856cb700-c1e8-11eb-8f57-1e4e20aeb127.png)

**Comment by bwikbs** [comment]
@xuelian-bai @swift-kim Any comments?

**Comment by xuelian-bai** [comment]
I'm not sure about this, but I think, if we want to support multi-screens, we can just create multiple flutter engines, each engine has its own window, this may consume many resources, but it's easy to do this way.  
I think multiple-flutters solution is actually from hybrid application on android(some pages are written in kotlin, some pages are written in dart). I will check more details tommorow.

**Comment by xuelian-bai** [comment]
for multiscreen requirement, is it for multiple monitor? One monitor display one window? Only one window can handle input events?  If that's the scenario, we can use eom to present another window(image) on another display. https://code.sec.samsung.net/confluence/display/WSI/EOM

**Comment by swift-kim** [comment]
@xuelian-bai Yes, I heard the second monitor wouldn't accept touch input and only display visual contents (mostly ads) without user interaction for now. Probably the API you mentioned will be used (I don't know the details). However, the scenario may be changed in the future and I think @bwikbs is preparing for possible future requirements.

**Comment by bwikbs** [comment]
> is it for multiple monitor?

yes

> One monitor display one window? Only one window can handle input events?

Nothing has been confirmed yet. I wish I want to know too.. :smile: 

``eom`` is the solution that other teams have been telling me to use for this. So we will probably use it. But I'm not sure if we can guide this to flutter app developers. 

There are many ways that how app developers use this, and I hope you will focus on that and share and discuss our ideas.

**Comment by xuelian-bai** [comment]
We may refer to this plugin https://github.com/VNAPNIC/presentation-displays.git

**Comment by xuelian-bai** [comment]
Here is my initial idea, didn't consider much details
 ![avatar]([https://github.com/xuelian-bai/rpi-vk-driver/blob/master/draft.png)

**Comment by bwikbs** [comment]
Thanks for giving me a good idea!
I have a question. 
Is this the same process? 
Could you explain a bit more running the engine? I have no idea where the dart is and where the native is.

**Comment by xuelian-bai** [comment]
it's just a rough process, here are more details, almost the same as the referred plugin:
dart app --> define external widget, main widget build() --> init eom plugin --> get outputid from eom --> create new flutter engine --> set window to outputid --> set initialroute as defined external widget for flutter engine in dart.
The referred plugin can use flutterview, but we can't, since we don't need to handle events for window on external display, I think it's ok to create two windows using eom, but we need to confirm this.

**Comment by xuelian-bai** [comment]
https://github.com/xuelian-bai/rpi-vk-driver/blob/master/draft2.png?raw=true

**Comment by xuelian-bai** [comment]
![avatar](https://github.com/xuelian-bai/rpi-vk-driver/blob/master/draft2.png?raw=true)

**Comment by bwikbs** [comment]
> The referred plugin can use flutterview, but we can't, since we don't need to handle events for window on external display, I think it's ok to create two windows using eom, but we need to confirm this.

Unfortunately no one can do this. :smile: 


-----
I think your approach is pretty good.(Especially,app developer doesn't need a native app interface) but still need a new engine instance and window instance.(Am I right?)
``FlutterEngineGroup`` API has resource and performance advantages. so.. I think it is still necessary to apply this to Tizen.

IMO, Now is the time to prepare for the unclear requirement, If we can afford it, it would be meaningful to POC both  and integrate them later.

**Comment by xuelian-bai** [comment]
> I think your approach is pretty good.(Especially,app developer doesn't need a native app interface) but still need a new engine instance and window instance.(Am I right?)

Yes, it needs a new engine instance, and I'm not sure if this engine instance can access "route" in main widget. 
We need a new window because eom can only set window to external display.

> `FlutterEngineGroup` API has resource and performance advantages. so.. I think it is still necessary to apply this to Tizen.

I checked android patch, lots of work, when will you start?

> IMO, Now is the time to prepare for the unclear requirement, If we can afford it, it would be meaningful to POC both and integrate them later.

I will assign xiaowei to investigate and implement eom related plugin, is that ok?

Actually I think there is another easier approach for the requirement, but it's kind of work around
1. Start app as usual
2. Implement the UI part on external display as offscreen widget, and make it as image
3. Create efl window inside app, attach the image to efl window, set efl window to eom output.
In this way, we only need one engine.

**Comment by bwikbs** [comment]
> I checked android patch, lots of work, when will you start?

Yes, you're right! And I haven't figured out what to do yet. My current plan is to create an API that uses the shell's spawn API first, and then slowly add the rest of the work and what we need. 

After the word came out, Let me share what I've checked so far...
Currently, When our embedder uses two engines at the same time, the only problem is the vsync waiter. This is not a problem with our implementation, but rather a problem that the upstream is not supposed to use two at same time.

Therefore, no special modifications are required at this time, and we can test 2 instances by adding spawn api to the embedder part.

> I will assign xiaowei to investigate and implement eom related plugin, is that ok?

Maybe @haesik  will answer this. :smile: 

> Actually I think there is another easier approach for the requirement, but it's kind of work around

If we don't have time and the scenario is finalized, we can consider it...But we still have time, so let's leave it as the last resort. :smiley:

**Comment by xuelian-bai** [comment]
> Currently, When our embedder uses two engines at the same time, the only problem is the vsync waiter. This is not a problem with our implementation, but rather a problem that the upstream is not supposed to use two at same time.
> 

Why two vsync are not supposed to use at the same time? I didn't check code of this part, but theoretically, If there are two engines, then there are two vsync, they are in different thread, why can't co-exist?

**Comment by haesik** [comment]
@xuelian-bai You can make the EOM related plugin of course, but we can change the direction or plan to support multiple screens

**Comment by bwikbs** [comment]
> we found that two engines are needed for eom plugin
yes, if only one engine with two windows, performance won't be good

@xuelian-bai Can you tell me something about this?

**Comment by bwikbs** [comment]
I think I've shared what I'm trying to do.
@swift-kim @xuelian-bai How do I handle below commit?
https://github.com/flutter-tizen/flutter-tizen/pull/126, https://github.com/flutter-tizen/engine/pull/98

IMO, there are two options.
1. Working on a separate dev branch.
2. Put it in main as an experimental feature.

Plz, give me your opinion.

**Comment by swift-kim** [comment]
@bwikbs If the design and API haven't been fixed yet, the first option looks better for me. If not many API changes will be made in the future, you might also consider the second option.

**Comment by bwikbs** [comment]
Thanks for your comment. I think the first one is better too.

**Comment by bwikbs** [comment]
**update**
I tested the shell's spawn API. It seems that simply using the API does not help with memory management.
(Except for the shell API, everything else is the same.)
1) Use shell's ``Spawn`` API
```
PID  S(CODE)  S(DATA)  P(CODE)  P(DATA)     PEAK      PSS       3D GEM(PSS) GEM(RSS)     SWAP  oom_score_adj    COMMAND
8957    73560        4    10032   111512   195108   152094     8552     1224     2448        0            200    /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner
```
2) Use shell's ``Create`` API
```
PID  S(CODE)  S(DATA)  P(CODE)  P(DATA)     PEAK      PSS       3D GEM(PSS) GEM(RSS)     SWAP  oom_score_adj    COMMAND
3731    73676        4     9976   111468   195124   152050     8624     1632     3264        0            200    /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner
```

**Comment by xuelian-bai** [comment]
@bwikbs did you run in release mode?

**Comment by bwikbs** [comment]
No, only debug mode. Is there any difference?

**Comment by xuelian-bai** [comment]
not sure if it's related, but you may see in function DartIsolate::CreateRootIsolate(), it will also check DartVM::IsRunningPrecompiledCode() if spawning_isolate. anyway, worth a try..

**Comment by bwikbs** [comment]
> not sure if it's related, but you may see in function DartIsolate::CreateRootIsolate(), it will also check DartVM::IsRunningPrecompiledCode() if spawning_isolate. anyway, worth a try.


Here is result of ``release``
As you said, the results is little different. Thanks for your tip! :+1: 
There seems to be some gain at ``PSS``, and the launch seems to be a little bit quicker.(by human sensor! :smile:   )
I feel it's worth moving on.


--- 

1. Use shell's ``Spawn`` API

| PID | S(CODE) | S(DATA) | P(CODE) | P(DATA) |PEAK |  PSS  | 3D GEM(PSS)| GEM(RSS) | SWAP | oom_score_adj |   COMMAND|
| -- | ---- | ---- |---- |---- |---- |---- |---- |---- |---- |---- |---- |
|9105  | 13200  |  4  |  8400 | 10736  | 32340 | 20151 | 8520 | 1632  | 3264 |  0 |  200 | /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner|
|9973  |  16544 |       4  |   8408  |  12020  |  36976  |  22448  |   8552   |  1224  |   2448  | 0 |  200 |   /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner|
|10636 |   16544  |      4   |  8408  |  11976  |  36932  |  22404  |   8552   |  1224  |   2448  | 0 |  200 |   /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner|



---------
2. Use shell's ``Create`` API

| PID | S(CODE) | S(DATA) | P(CODE) | P(DATA) |PEAK |  PSS  | 3D GEM(PSS)| GEM(RSS) | SWAP | oom_score_adj |   COMMAND|
| -- | ---- | ---- |---- |---- |---- |---- |---- |---- |---- |---- |---- |
|9104 | 22140 | 4 |  5600 | 13632 | 41376 | 24047 | 8624 | 2040 |  4080 | 0 |   200 | /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner|
|10853 | 22132 |  4  | 5600 |  13612  |  41348 |   24027  |   8624  |   2040  |   4080   |   0 |   200  |  /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner|
|11495 | 20768 |  4  | 6964 |  13680  |  41416 |   24973  |   8688  |   2040  |   4080   |   0 |   200  |  /opt/usr/globalapps/com.example.multiple_flutters_tizen/bin/runner|

**Comment by xuelian-bai** [comment]
hmm. launch is definitely quicker, as I tested, release mode will decrease about 500ms when preparing dart isolate, since it's using AOT.

**Comment by xiaowei-guan** [comment]
https://github.com/xiaowei-guan/engine/tree/dart_entrypoint
I have finished sample code for multiple screens based on EOM in flutter engine.
if we want to full support multiple screens, we need :

1. support dart entrypoint.
2. support engine group.
3. fix vsync issue

**Comment by bwikbs** [comment]
Oh, thank you! 
If things go well, it might be possible to run an [example](https://github.com/flutter/samples/tree/master/add_to_app/multiple_flutters) by referring to this.

**Comment by JSUYA** [comment]
https://github.com/flutter-tizen/embedder/pull/3
A patch to support EOM has been merged. Note please.
